[#413] Form elements in disabled fieldsets are not tabbable/focusable

stefcameron opened the pull request:
Fixes #413

<!--
Thank you for your contribution! 🎉

Please be sure to go over the PR CHECKLIST below before posting your PR to make sure we all think of "everything". :)
-->

<details>
<summary>PR Checklist</summary>
<br/>

__Please leave this checklist in your PR.__

- Source changes maintain stated browser compatibility.
- Issue being fixed is referenced.
- Unit test coverage added/updated.
- E2E test coverage added/updated.
- Typings added/updated.
- README updated (API changes, instructions, etc.).
- Changes to dependencies explained.
- Changeset added (run `yarn changeset` locally to add one, and follow the prompts).
  - EXCEPTION: A Changeset is not required if the change does not affect any of the source files that produce the package bundle. For example, tooling changes, test updates, or a new dev-only dependency to run tests more efficiently should not have a Changeset since it will not affect package consumers.

</details>

stefcameron commented:
@MartijnCuppens FYI. If you have a moment, please check it out and see if it covers your use case. Maybe there's some other form element I'm not thinking of, or some other scenario I'm not thinking of...?

MartijnCuppens commented:
> @MartijnCuppens FYI. If you have a moment, please check it out and see if it covers your use case. Maybe there's some other form element I'm not thinking of, or some other scenario I'm not thinking of...?

LGTM

I only bumped into an edge case while reading the specs:
> A form control is disabled if its disabled attribute is set, or if it is a descendant of a fieldset element whose disabled attribute is set and is not a descendant of that fieldset element's first legend element child, if any.
> 
> Source: https://www.w3.org/TR/2014/REC-html5-20141028/forms.html#enabling-and-disabling-form-controls:-the-disabled-attribute

I've updated the codepen with some extra edge cases: https://codepen.io/MartijnCuppens/pen/abWRZNY?editors=1011

stefcameron commented on src/index.js:
```diff
@@ -165,13 +165,36 @@ const isHidden = function (node, displayCheck) {
+    node.tagName === 'BUTTON'
+  ) {
+    let parentNode = node.parentElement;
+    while (parentNode) {
```
My first thought was to use a query, but I don't think there's a selector that can do this, especially because the disabled fieldset could be one that is an ancestor far removed from where the node actually is in the hierarchy within it.

idoros replied:
Fair enough. :where selector would have been nice

stefcameron replied:
Let's say we used this as the selector, `fieldset:not([disabled]) :where(input, button, textarea, select)` It still selects an input inside an enabled fieldset, which is nested inside a disabled fieldset -- and the fact the parent fieldset is disabled means that input will be disabled, even though it's nearest parent is an enabled fieldset.

idoros replied:
without taking the `<legend>` rule this would have targeted all buttons that have no `[disabled]` attribute and aren't nested under a `[disabled]` `<fieldset>`:

```css
button:not(
  [disabled],
  :where(fieldset[disabled] *)
)
```

adding buttons that are nested under the first `<legend>` but not under a any other more nested `<legend>` is more difficult.

idoros replied:
yes, this is difficult :(

stefcameron commented:
@MartijnCuppens
> > @MartijnCuppens FYI. If you have a moment, please check it out and see if it covers your use case. Maybe there's some other form element I'm not thinking of, or some other scenario I'm not thinking of...?
> 
> LGTM
> 
> I only bumped into an edge case while reading the specs:
> 
> > A form control is disabled if its disabled attribute is set, or if it is a descendant of a fieldset element whose disabled attribute is set and is not a descendant of that fieldset element's first legend element child, if any.
> > Source: https://www.w3.org/TR/2014/REC-html5-20141028/forms.html#enabling-and-disabling-form-controls:-the-disabled-attribute
> 
> I've updated the codepen with some extra edge cases: https://codepen.io/MartijnCuppens/pen/abWRZNY?editors=1011

Thank you, that was very helpful! I've updated the code to support the legend corner cases.

MartijnCuppens commented on src/index.js:
```diff
@@ -165,13 +165,58 @@ const isHidden = function (node, displayCheck) {
+        //  (however deep nested) and it'll be disabled
+        for (let i = 0; i < parentNode.children.length; i++) {
+          const child = parentNode.children.item(i);
+          if (child.tagName === 'LEGEND') {
```
I don't think it is necessary to loop over all elements since it's only valid to use the legend as first element of the fieldset:
https://html.spec.whatwg.org/multipage/form-elements.html#the-legend-element

Apart from that minor thing, this looks good!

stefcameron replied:
I don't like the loop, but I don't know how else to check. Note, however, that this is only looping over immediate children of the first-found disabled fieldset up the hierarchy from the node in question, and then looking for the first-found legend as an immediate (not deep) child of that fieldset to check (now that may be quite deep, but how else to check without going deep? hopefully `contains` is efficient in that regard) if the node in question is a child of it. Early breaks in all cases.

MartijnCuppens replied:
You're right, I can't find a more performant way either!

stefcameron replied:
Thanks for checking it out!
